core/dataset.cpp

```diff
@@ -651,6 +665,70 @@ decltype(auto) apply_with_delay(const Op &op, A &&a, const B &b) {
   return std::forward<A>(a);
 }
 
+template <class T> void copy_metadata(Dataset &dest, const T &src) {
+  /* Dense coordinates */
+  for (const auto & [ name, value ] : src.coords()) {
+    dest.setCoord(name, value);
+  }
+
+  /* Dense labels */
+  for (const auto & [ name, value ] : src.labels()) {
+    dest.setLabels(std::string(name), value);
+  }
+
+  /* Attributes */
+  for (const auto & [ name, value ] : src.attrs()) {
+    dest.setAttr(std::string(name), value);
+  }
+}
+
+void copy_metadata(Dataset &dest, const std::string &name,
+                   const DataConstProxy &src) {
+  /* Sparse coordinates */
+  for (const auto &coord : src.coords()) {
+    if (coord.second.dims().sparse()) {
+      dest.setSparseCoord(name, coord.second);
+    }
+  }
+
+  /* Sparse labels */
+  for (const auto & [ label_name, labels ] : src.labels()) {
+    if (labels.dims().sparse()) {
+      dest.setSparseLabels(name, std::string(label_name), labels);
+    }
+  }
+}
+
+template <class Op, class A, class B>
+auto apply_with_broadcast(const Op &op, const A &a, const B &b) {
+  expect::coordsAndLabelsMatch(a, b);
+
+  Dataset res;
+  copy_metadata(res, a);
+
+  for (const auto & [ name, item ] : b) {
+    if (a.contains(name)) {
+      res.setData(std::string(name), op(a[name].data(), item.data()));
+      copy_metadata(res, std::string(name), a[name]);
```

Comment:
It can happen that `a` is dense and `b` is sparse. The we would need to `copy_metadata` from `b`.

Reply:
Would we actually want to support that? What would happen in that case?

Reply:
Yes, for example to add a constant offset to all values.

Reply:
Or to get the inverse:

```cpp
auto inverse = 1.0/dataset
```